tests/phpunit/tests/dependencies/scripts.php

```diff
@@ -586,6 +586,105 @@ public function test_get_normalized_script_args() {
+	 *
+	 * @ticket 12009
+	 */
+	public function test_concatenate_with_defer_strategy() {
+		global $wp_scripts, $concatenate_scripts;
+
+		$old_value = $concatenate_scripts;
+		$concatenate_scripts = true;
+
+		$wp_scripts->do_concat    = true;
+		$wp_scripts->default_dirs = array( '/directory/' );
+
+		wp_register_script( 'one-concat-dep', '/directory/script.js' );
+		wp_register_script( 'two-concat-dep', '/directory/script.js' );
+		wp_register_script( 'three-concat-dep', '/directory/script.js' );
+		wp_enqueue_script( 'main-defer-script', '/main-script.js', array( 'one-concat-dep', 'two-concat-dep', 'three-concat-dep' ), null, array( 'strategy' => 'defer' ) );
+
+		wp_print_scripts();
+		$print_scripts = get_echo( '_print_scripts' );
+
+		// reset global before asserting.
+		$concatenate_scripts = $old_value;
+
+		$ver       = get_bloginfo( 'version' );
+		$expected  = "<script type='text/javascript' src='/wp-admin/load-scripts.php?c=0&amp;load%5Bchunk_0%5D=one-concat-dep,two-concat-dep,three-concat-dep&amp;ver={$ver}'></script>\n";
+		$expected .= "<script type='text/javascript' src='/main-script.js' id='main-defer-script-js' defer></script>\n";
+
+		$this->assertSame( $expected, $print_scripts );
+	}
+
+	/**
+	 * Test script concatenation with `async` main script.
+	 *
+	 * @ticket 12009
+	 */
+	public function test_concatenate_with_async_strategy() {
+		global $wp_scripts, $concatenate_scripts;
+
+		$old_value = $concatenate_scripts;
+		$concatenate_scripts = true;
+
+		$wp_scripts->do_concat    = true;
+		$wp_scripts->default_dirs = array( '/directory/' );
+
+		wp_enqueue_script( 'one-concat-dep-1', '/directory/script.js' );
+		wp_enqueue_script( 'two-concat-dep-1', '/directory/script.js' );
+		wp_enqueue_script( 'three-concat-dep-1', '/directory/script.js' );
+		wp_enqueue_script( 'main-defer-script-1', '/main-script.js', array(), null, array( 'strategy' => 'async' ) );
+
+		wp_print_scripts();
+		$print_scripts = get_echo( '_print_scripts' );
+
+		// reset global before asserting.
+		$concatenate_scripts = $old_value;
+
+		$ver       = get_bloginfo( 'version' );
+		$expected  = "<script type='text/javascript' src='/wp-admin/load-scripts.php?c=0&amp;load%5Bchunk_0%5D=one-concat-dep-1,two-concat-dep-1,three-concat-dep-1&amp;ver={$ver}'></script>\n";
+		$expected .= "<script type='text/javascript' src='/main-script.js' id='main-defer-script-1-js' async></script>\n";
+
+		$this->assertSame( $expected, $print_scripts );
+	}
+
+	/**
+	 * Test script concatenation with blocking scripts before and after a `defer` script. 
+	 *
+	 * @ticket 12009
+	 */
+	public function test_concatenate_with_blocking_script_before_and_after_script_with_defer_strategy() {
+		global $wp_scripts, $concatenate_scripts;
+
+		$old_value = $concatenate_scripts;
+		$concatenate_scripts = true;
+
+		$wp_scripts->do_concat    = true;
+		$wp_scripts->default_dirs = array( '/directory/' );
+
+		wp_enqueue_script( 'one-concat-dep-2', '/directory/script.js' );
```

Comment:
This makes me curious about what we would expect to happen if there's a mix of concatenated scripts that depend on a deferred script or if only some of these concatenated scripts were dependencies of the deferred script.

Reply:
If any script depends on a deferred script, the deferred script itself will be loaded in a blocking way. `get_strategy` takes care of that.